### PR TITLE
36 python idw

### PR DESCRIPTION
**Pull request by rosepearson**
Switch from pdal.writers.gdal IDW to a python implementation based on the scipy.spatial.KDTree. See [issue IDW](https://github.com/rosepearson/GeoFabrics/issues/36) for the reasons behind this change. Mainly they centre around preparation for using Dask. Benchmarking has been done in this [document](https://niwa.sharepoint.com/:w:/r/sites/FloodResilienceAotearoa/NationalFloodMapping/Shared%20Documents/DEM%20Generation/Benchmarking%20processing%20times.docx?d=w8873a810f7af4f8e844291369b004099&csf=1&web=1&e=ZQq1hw).
 - [x] Make code change
 - [x] Update tests 
   - [x] Update / create new tests
   - [x] Ensure these tests have the expected behavour
   - [x] Test locally and ensure tests are passing
 - [x] Ensure tests passing on GH Actions
 - [x] Update documentation
   - [x] Doc strings
   - [x] Wiki 
  - [x] Update package version (if needed)

### CODE REVIEW

**Comment by rosepearson** [comment]
Hi @AliceHarang, 
This is ready for a review now. The main changes are in dem.py. I have changed the IDW implementation used to perform IDW interpolation where there is LiDAR data.

There are also associated changes to the tests - benchmark files have been updated and the Linux test failures have been updated. 

As always, let me know if you have any questions!

File: src/geofabrics/processor.py [hunk 1]
```diff
@@ -219,9 +219,16 @@ def get_lidar_dataset_crs(self, data_service, dataset_name) -> dict:
             return None
 
     def get_lidar_file_list(self, data_service) -> dict:
-        """ Return a dictionary that contains a list of LiDAR tiles to process under the key 'file_paths' and optionally
-        the CRS of these files under the key 'crs'. The 'crs' should only be set if the CRS information is not correctly
-        encoded in the LAZ/LAS files.
+        """ Return a dictionary with three enties 'file_paths', 'crs' and 'tile_index_file'. The 'file_paths' contains a
+        list of LiDAR tiles to process.
+
+        The 'crs' is only optionally set (if unset the value is None). The 'crs' should only be set if the CRS
```
Comment (AliceHarang) [inline src/geofabrics/processor.py:225]: What crs mean?

Comment (rosepearson) [reply src/geofabrics/processor.py:225]: I'll copy in the longer comment from elsewhere. Sorry I should have included it in this docstring. CRS effectively means the coordinate system

Comment (rosepearson) [reply src/geofabrics/processor.py:225]: I have changed to "The 'crs' (or coordinate system of the LiDAR data as defined by an EPSG code) is only optionally set (if unset the value is None). The 'crs' should only be set if the CRS information is not correctly encoded in the LAZ/LAS  files. Currently this is only supported for OpenTopography LiDAR." and will push this commit along with any other edits you pick-up.

File: src/geofabrics/lidar.py [hunk 2]
```diff
@@ -15,24 +15,33 @@
 
 
 class CatchmentLidar:
-    """ A class to manage LiDAR data in a catchment context by supporting the addition of LiDAR data tile by tile.
+    """ A class to manage LiDAR data in a catchment context by supporting the addition of LiDAR data tile by tile. LiDAR
+    files are loaded using PDAL and the point cloud, and it's extents are returned.
     """
     LAS_GROUND = 2
```
Comment (AliceHarang) [inline src/geofabrics/lidar.py:21]: Can this value change depending of the type of data?

**Review by AliceHarang — APPROVED** [review]
Hi Rose,
I think that this implementation is useful (wait to see the library working in parallel !) and well explain!

File: src/geofabrics/dem.py [hunk 3]
```diff
@@ -329,205 +324,213 @@ def _set_up(self, catchment_geometry):
...
+        z_out = numpy.zeros(len(xy_out))
+
+        for i, (near_indicies, point) in enumerate(zip(tree_index_list, xy_out)):
+
+            if len(near_indicies) == 0:  # Set NaN if no values in search region
+                z_out[i] = numpy.nan
+            else:
+                distance_vectors = point - tree.data[near_indicies]
+                smoothed_distances = numpy.sqrt(((distance_vectors**2).sum(axis=1)+smoothing**2))
+                if smoothed_distances.min() == 0:  # incase the of an exact match
```
[the remark this answers is not in the document]
Comment (rosepearson) [reply src/geofabrics/dem.py:366]: Haha - I read this about 10 times before I saw it. My mind new what I was trying to say at least :)

File: src/geofabrics/dem.py [hunk 4]
```diff
@@ -329,205 +324,213 @@ def _set_up(self, catchment_geometry):
...
-
-        # Assert PDAL wrote the DEm tile to the temp file name
-        metadata = json.loads(pdal_pipeline.get_metadata())
-        assert str(self._temp_dem_file) == metadata['metadata']['writers.gdal']['filename'][0], "The DEM tile has " + \
-            " been written out in an unexpected location. It has been witten out to " + \
-            f"{metadata['metadata']['writers.gdal']['filename'][0]} instead of {self._temp_dem_file}"
+        return z_out
 
     def add_tile(self, tile_points: numpy.ndarray, tile_extent: geopandas.GeoDataFrame, window_size: int,
                  idw_power: int, radius: float, method: str = 'first'):
```
[the remark this answers is not in the document]
Comment (rosepearson) [reply src/geofabrics/dem.py:375]: Well googled. I've clarified here and elsewhere!

File: src/geofabrics/lidar.py (hunk 2 above)
Comment (rosepearson) [reply src/geofabrics/lidar.py:21]: No this is specified in the LAS/LAZ format. I'll add a comment to that effect :)

File: src/geofabrics/dem.py [hunk 5]
```diff
@@ -329,205 +324,213 @@ def _set_up(self, catchment_geometry):
...
+                polygon_out = geopandas.GeoDataFrame({'geometry': [polygon]},
+                                                     crs=self.catchment_geometry.crs['horizontal'])
+                polygon_out = geopandas.clip(self.catchment_geometry.catchment, polygon_out)
+            else:
+                if self.verbose:
+                    print("Warning filtering holes in CatchmentLidar using filter_lidar_extents_for_holes is not yet "
+                          + f"supported for {polygon.geometryType()}")
+        return polygon_out
+
+    def _filter_holes_around_polygon(self, area_to_filter: float,
```
[the remark this answers is not in the document]
Comment (rosepearson) [reply src/geofabrics/dem.py:503]: Cool - I have added some explanatory text.

File: src/geofabrics/dem.py [hunk 6]
```diff
@@ -329,205 +324,213 @@ def _set_up(self, catchment_geometry):
...
+        dim_y = numpy.arange(catchment_bounds[3] - resolution / 2, catchment_bounds[1], -resolution,
+                             dtype=self.raster_type)
+        grid_dem_z = numpy.empty((1, len(dim_y), len(dim_x)), dtype=self.raster_type)
+        dem = xarray.DataArray(grid_dem_z, coords={'band': [1], 'y': dim_y, 'x': dim_x}, dims=['band', 'y', 'x'],
+                               attrs={'scale_factor': 1.0, 'add_offset': 0.0, 'long_name': 'idw'})
+        dem.rio.write_crs(catchment_geometry.crs['horizontal'], inplace=True)
+        dem.name = 'z'
+        dem = dem.rio.write_nodata(numpy.nan)
+
+        # Ensure the DEM is empty - all NaN - and clipped to size
```
[the remark this answers is not in the document]
Comment (rosepearson) [reply src/geofabrics/dem.py:341]: This is actually to initialise the array as all NaN. These values are only set to something other later as part of the IDW and linear RBF interpolation.

**Pull request merged**